x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/huggingface/HuggingFaceElserAction.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.external.action.huggingface;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.ElasticsearchStatusException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.inference.InferenceResults;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.xpack.inference.external.action.ExecutableAction;
+import org.elasticsearch.xpack.inference.external.http.sender.Sender;
+import org.elasticsearch.xpack.inference.external.huggingface.HuggingFaceAccount;
+import org.elasticsearch.xpack.inference.external.huggingface.HuggingFaceClient;
+import org.elasticsearch.xpack.inference.external.request.huggingface.HuggingFaceElserRequest;
+import org.elasticsearch.xpack.inference.external.request.huggingface.HuggingFaceElserRequestEntity;
+import org.elasticsearch.xpack.inference.services.huggingface.elser.HuggingFaceElserSecretSettings;
+import org.elasticsearch.xpack.inference.services.huggingface.elser.HuggingFaceElserServiceSettings;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import static org.elasticsearch.core.Strings.format;
+
+public class HuggingFaceElserAction implements ExecutableAction {
+    private final String input;
+    private final Sender sender;
+    private final HuggingFaceElserServiceSettings serviceSettings;
+    private final HuggingFaceElserSecretSettings secretSettings;
+
+    public HuggingFaceElserAction(
+        String input,
+        Sender sender,
+        HuggingFaceElserServiceSettings serviceSettings,
+        HuggingFaceElserSecretSettings secretSettings
+    ) {
+        this.input = input;
+        this.sender = sender;
+        this.serviceSettings = serviceSettings;
+        this.secretSettings = secretSettings;
+    }
+
+    public void execute(ActionListener<InferenceResults> listener) {
+        try {
+            URI url = new URI(serviceSettings.url());
+            HuggingFaceAccount account = new HuggingFaceAccount(url, secretSettings.apiKey());
+            HuggingFaceElserRequestEntity entity = new HuggingFaceElserRequestEntity(input);
+            HuggingFaceElserRequest request = new HuggingFaceElserRequest(account, entity);
+            HuggingFaceClient client = new HuggingFaceClient(sender);
```

Comment:
Good point, I'll move them to the constructor. At the moment the `execute` method of these objects is only ever called once per object construction.

Reply:
Actually, if we keep the `URI` construction in `execute` then we can limit the `try/catch` to only this class I think. Otherwise I think we'd need to `try/catch` around `HuggingFaceElserAction` because it could throw a URI syntax exception and the around the `send` because of a possible `IOException`. I'll move the classes that don't rely on that.

Maybe it makes more sense to pass in the `listener` to the constructor 🤔

Reply:
I don't think you can move the `listener` to the constructor. That's likely a lambda that's unique per inference call.

But `url`, `account` and `client` could just be constructed once.

> we can limit the `try/catch` to only this class

But doesn't that mean that any problems with basic construction of `url`, `account` or `client` will result in an error per call rather than a single error at startup? That's not ideal. It would be best if such basic errors could be reported when the service is first created.